src/js/components/Button/Button.js

```diff
@@ -92,12 +96,80 @@ const getIconColor = (paths = [], theme, colorProp, kind) => {
   return result[1] || undefined;
 };
 
+const getBadgeDimension = (dimension, badgeProp, badgeRef, theme) => {
+  if (
+    typeof badgeProp === 'number' ||
+    (typeof badgeProp === 'object' && badgeProp.value)
+  ) {
+    const borderWidth = badgeProp.border
+      ? parseInt(
+          theme.global.borderSize[badgeProp.border.size].replace('px', ''),
+          10,
+        ) * 2
+      : 0;
+    // leave a small amount of horizontal space to pad content
+    const horizontalPad =
+      dimension === 'width'
+        ? parseInt(theme.global.edgeSize.xsmall.replace('px', ''), 10)
+        : 0;
+    // if content is tall/wide, let badge grow to fit. otherwise,
+    // make sure it's at least badge.size.medium dimensions
+    return `${Math.max(
+      Math.ceil(badgeRef.current.getBoundingClientRect()[dimension]) +
+        horizontalPad +
+        borderWidth,
+      parseInt(theme.button.badge.size.medium.replace('px', ''), 10) +
+        borderWidth,
+    )}px`;
+  }
+  return `${badgeRef.current.getBoundingClientRect()[dimension]}px`;
+};
+
+const getBadge = (badgeProp, badgeRef, badgeHeight, badgeWidth, theme) => {
+  const max = badgeProp.max || theme.button.badge.max;
+
+  let value;
+  if (typeof badgeProp === 'number') value = badgeProp;
+  else if (typeof badgeProp === 'object') value = badgeProp.value;
+  let badge;
+  if (
+    typeof value === 'number' ||
+    typeof value === 'boolean' ||
+    typeof badgeProp === 'boolean'
+  ) {
+    if (typeof value === 'number') {
+      badge = (
+        <Text color="text-strong" size="small" weight="normal" ref={badgeRef}>
+          {value > max ? `${max}+` : value}
+        </Text>
+      );
+    }
+    badge = (
+      <Box
+        align="center"
+        background={badgeProp.background || theme.button.badge.background}
+        border={badgeProp.border || theme.button.badge.border}
+        flex={false}
+        height={badgeHeight}
+        justify="center"
+        round
+        width={badgeWidth}
+      >
+        {badge}
+      </Box>
+    );
+    // caller has provided their own JSX and we will just render that
+  } else badge = <Box ref={badgeRef}>{badgeProp}</Box>;
```

Comment:
Something seems inconsistent that `badgeRef` here refers to the whole badge but when teh value is a number, it's just the text content.

Reply:
Initially, I intended to clone the badgeProp and pass the ref directly onto the badgeProp (which would be more aligned with the concept of "content", but I was running into issue where the badgeProp would need to be built using `forwardRef` to properly receive the ref and didn't think it would be reasonable to except the caller to implement with that. For that reason, I wrapped the badgeProp in a box.